src/model/formula/Formula.Clause.js

```diff
@@ -34,40 +33,73 @@ class Clause extends Formula {
     }
     this.litsMultiset = [];
     var contains = false;
-    for (let lit in this.lits){
+    for (let lit of this.lits){
       contains = false;
-      for (let tuple in this.litsMultiset){
-        if (tuple[0] === lit.name) {
+      for (let tuple of this.litsMultiset){
+        if (tuple[0].equals(lit)) {
           tuple[1]++;
           contains = true;
           break;
         }
       }
       if (!contains) {
-        this.litsMultiset.push([lit.name, 1]);
+        this.litsMultiset.push([lit, 1]);
       }
     }
-    return this.litsMultiset = [];
+    return this.litsMultiset;
   }
   
   has(key){
-    for (let tuple of this.litsMultiset) {
-      if (tuple[0] === key){
+    for (let tuple of this.getLitsMultiset()) {
+      if (tuple[0].equals(key)){
         return true;
       }
     }
     return false;
   }
 
   get(key){
-    for (let tuple of this.lits) {
-      if (tuple[0] === key){
+    for (let tuple of this.getLitsMultiset()) {
+      if (tuple[0].equals(key)){
         return tuple[1];
       }
     }
     return undefined;
   }
 
+  resolution(cl1, cl2, sub){
+    let newCl1 = cl1.substitute(sub);
+    let newCl2 = cl2.substitute(sub);
+    for (let lit1 of newCl1.lits){
+      for (let lit2 of newCl2.lits){
+        if (lit1.equals(lit2.negation())){
+          if (this.equals(new Clause(newCl1.lits.filter(lit => !lit.equals(lit1))
+            .concat(newCl2.lits.filter(lit => !lit.equals(lit2)))))){
+            return true;
+          }
+          break;
+        }
+      }
+    }
+    return false;
+  }
+
+  factorisation(cl, sub){
+    let newCl = cl.substitute(sub);
+    for (let i = 0; i < newCl.lits.length; i++){
+      for (let j = i+1; j < newCl.lits.length; j++){
+        if (newCl.lits[i].equals(newCl.lits[j])){
+          if (this.equals(new Clause(newCl.lits.filter(lit => !lit.equals(newCl.lits[i]))
+            .concat([newCl.lits[i]])))) {

```

Comment:
Faktorizacia by mala fungovat tak, ze ak po unifikacii dostaneme klauzulu ` L ∨ K ∨ L ∨ M ∨ L`, faktorom je klauzula, ktorá ma o jeden vyskyt `L` menej, teda ` L ∨ K ∨ L ∨ M`. Vasa `new Clause()` bude ale obsahovat iba jeden vyskyt `L`.

Riesenie je jednoduche a navyse zase zefektivni vnorene cykly: Staci odfiltrovat `i`-ty literal  a nic dalsie do klauzuly nepridavat: `const factor = new Clause(newCl.lits.filter((_, k) => i !== k))`. Vsimnite si, ze `factor` nezavisi od `j`, takze ho mozete vyrobit uz vo *vonkajsom* cykle a hned aj porovnat s `this`. **Ak** su rovnake, vo vnutornom cykle sa pokusite najst `j`-ty literal v `newCl` rovny `i`-temu a ak existuje, vratite `true`. Inak alebo ak sa `this` nerovna `factor`, skusite dalsie `i`.